### PR TITLE
Use PHPStan Scope over custom Reflection and own type resolution

### LINKED ISSUE #497: Add Broker for class types
Broker for classes, for method

- store all info there, to fetch them just once
- for reflectoin

inspired by: https://github.com/phpstan/phpstan/blob/3acb1eebcb173bfbbb43acc1a029fd544d10287c/src/Broker/Broker.php

kinda alternative to BetterReflection - use cache wrapper over that instead?

### PR DESCRIPTION
**Pull request by TomasVotruba**
- Closes #497

### Related Changes

- drop `Expr` node analysis from `NodeTypeResolver` and let PHPStan handle them - https://github.com/rectorphp/rector/pull/544/commits/28feececa100bdf9e32946db384b888b836926dd
- add `Broker` service and `Scope` objects to Rector DI Container 

### Todo

- [x] remove SmartClassReflector
- [x] remove BetterReflection

### CODE REVIEW

File: packages/NodeTypeResolver/src/NodeVisitor/PHPStanScopeNodeVisitor.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * @return int|null|Node
+     */
+    public function enterNode(Node $node)
+    {
+        $this->nodeScopeResolver->setAnalysedFiles([$this->currentFileProvider->getSplFileInfo()->getRealPath()]);
+
+        // it needs to be done here, because all other NodeVisitors enterNode() methods have to be run
+        $this->nodeScopeResolver->processNodes([$node], $this->scope, function (Node $node, Scope $scope): void {
+            // Record scope
```
Comment (TomasVotruba) [inline packages/NodeTypeResolver/src/NodeVisitor/PHPStanScopeNodeVisitor.php]: Use all nodes not just one

https://github.com/phpstan/phpstan/blob/2139878e2aa79e908a6bf928157adec5d0ec377c/src/Analyser/Analyser.php#L123

Probably in standalone node traverser

**Comment by carusogabriel** [comment]
@TomasVotruba Will this help us in performance, is just a code clean up? :blush:

**Comment by TomasVotruba** [comment]
Both :) I think we can delegate 60-70 % of type resolving to PHPStan.

**Comment by TomasVotruba** [comment]
Just see this commit: https://github.com/rectorphp/rector/pull/544/commits/28feececa100bdf9e32946db384b888b836926dd

**Comment by carusogabriel** [comment]
@TomasVotruba Let me know when you finish, so I can review it before we merge :)

**Comment by TomasVotruba** [comment]
Ok, I will :+1:

**Comment by carusogabriel** [comment]
Gonna review it ASA I can, please, don't merge before it.

**Comment by TomasVotruba** [comment]
I have nothing to add or remove now. It's ready, feel free to merge if you agree

File: packages/NodeTypeResolver/src/PHPStan/Type/TypeToStringResolver.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    public function resolve(Type $type): array
+    {
+        $types = [];
+
+        if ($type instanceof ObjectType) {
+            $types[] = $type->getClassName();
+        }
+
+        if ($type instanceof UnionType || $type instanceof IntersectionType) {
+            foreach ($type->getTypes() as $type) {
```
Comment (carusogabriel) [inline packages/NodeTypeResolver/src/PHPStan/Type/TypeToStringResolver.php]: Maybe an early return here?

File: packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php [hunk 3]
```diff
@@ -42,10 +42,11 @@ public function dataProvider(): Iterator
 
         yield [__DIR__ . '/Source/ClassWithTrait.php', 0, [ClassWithTrait::class, AnotherTrait::class]];
 
+        // traits in anonymous classes are ignored in PHPStan
         yield [
             __DIR__ . '/Source/AnonymousClass.php',
             0,
-            [ParentClass::class, SomeInterface::class, AnotherTrait::class],
+            [ParentClass::class, SomeInterface::class /*, AnotherTrait::class */],
```
Comment (carusogabriel) [inline packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php]: Maybe remove this test?

File: src/config/services.yml [hunk 4]
```diff
@@ -12,15 +12,16 @@ services:
 
     # prepares helper attributes of nodes
     Rector\NodeTraverser\StandaloneTraverseNodeTraverser:
-        # order matters, so the next NodeVisitor can work with previous' attributes
+#        # order matters, so the next NodeVisitor can work with previous' attributes
```
Comment (carusogabriel) [inline src/config/services.yml]: Duplicated comments?

File: tests/Rector/Assign/PropertyAssignToMethodCallRector/PropertyAssignToMethodCallRectorTest.php [hunk 5]
```diff
@@ -20,7 +20,7 @@ public function test(string $wrong, string $fixed): void
 
     public function provideWrongToFixedFiles(): Iterator
     {
-        yield [__DIR__ . '/Wrong/wrong.php.inc', __DIR__ . '/Correct/correct.php.inc'];
+//        yield [__DIR__ . '/Wrong/wrong.php.inc', __DIR__ . '/Correct/correct.php.inc'];
```
Comment (carusogabriel) [inline tests/Rector/Assign/PropertyAssignToMethodCallRector/PropertyAssignToMethodCallRectorTest.php]: Dead test?

File: tests/Rector/Class_/ClassReplacerRector/ClassReplacerRectorTest.php [hunk 6]
```diff
@@ -20,9 +20,8 @@ public function test(string $wrong, string $fixed): void
 
     public function provideWrongToFixedFiles(): Iterator
     {
-        yield [__DIR__ . '/Wrong/wrong.php.inc', __DIR__ . '/Correct/correct.php.inc'];
+//        yield [__DIR__ . '/Wrong/wrong.php.inc', __DIR__ . '/Correct/correct.php.inc'];
```
Comment (carusogabriel) [inline tests/Rector/Class_/ClassReplacerRector/ClassReplacerRectorTest.php]: Dead test?

File: tests/Rector/Typehint/ParentTypehintedArgumentRector/ParentTypehintedArgumentRectorTest.php [hunk 7]
```diff
@@ -18,7 +21,7 @@ public function test(string $wrong, string $fixed): void
     public function provideWrongToFixedFiles(): Iterator
     {
         yield [__DIR__ . '/Wrong/wrong.php.inc', __DIR__ . '/Correct/correct.php.inc'];
-        yield [__DIR__ . '/Wrong/wrong2.php.inc', __DIR__ . '/Correct/correct2.php.inc'];
+//        yield [__DIR__ . '/Wrong/wrong2.php.inc', __DIR__ . '/Correct/correct2.php.inc'];
```
Comment (carusogabriel) [inline tests/Rector/Typehint/ParentTypehintedArgumentRector/ParentTypehintedArgumentRectorTest.php]: Dead test?

File: .travis.yml [hunk 8]
```diff
@@ -29,11 +29,10 @@ script:
   - |
     if [[ $RUN_RECTOR == true ]]; then
       bin/rector process src tests --level symfony40 --dry-run
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply .travis.yml]: This just simple check of rector runs with some level

File: packages/NodeTypeResolver/src/PHPStan/Type/TypeToStringResolver.php (hunk 2 above)
Comment (TomasVotruba) [reply packages/NodeTypeResolver/src/PHPStan/Type/TypeToStringResolver.php]: There will be more types checked in the future.

File: packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php (hunk 3 above)
Comment (TomasVotruba) [reply packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php]: Why?

File: src/config/services.yml (hunk 4 above)
Comment (TomasVotruba) [reply src/config/services.yml]: Yes, I'll fix it

File: tests/Rector/Assign/PropertyAssignToMethodCallRector/PropertyAssignToMethodCallRectorTest.php (hunk 5 above)
Comment (TomasVotruba) [reply tests/Rector/Assign/PropertyAssignToMethodCallRector/PropertyAssignToMethodCallRectorTest.php]: I'll check it

File: tests/Rector/Typehint/ParentTypehintedArgumentRector/ParentTypehintedArgumentRectorTest.php (hunk 7 above)
Comment (TomasVotruba) [reply tests/Rector/Typehint/ParentTypehintedArgumentRector/ParentTypehintedArgumentRectorTest.php]: I'll check it

File: tests/Rector/Class_/ClassReplacerRector/ClassReplacerRectorTest.php (hunk 6 above)
Comment (TomasVotruba) [reply tests/Rector/Class_/ClassReplacerRector/ClassReplacerRectorTest.php]: I'll check it

**Comment by TomasVotruba** [comment]
Thank you Gabriel for nice words and review. I see the value in less code to maintain as well, that will add more value by Rector in time.

I'll check PHPUnit test run as brenchmark.

**Comment by TomasVotruba** [comment]
### Run on master

```bash
PHPUnit 7.2.6 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.7-0ubuntu0.18.04.2
Configuration: /var/www/rector/phpunit.xml

...............................................................  63 / 227 ( 27%)
............................................................... 126 / 227 ( 55%)
............................................................... 189 / 227 ( 83%)
......................................                          227 / 227 (100%)

Time: 11.9 seconds, Memory: 360.00MB

OK (227 tests, 417 assertions)
```

### 1st run on this PR

```bash
PHPUnit 7.2.6 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.7-0ubuntu0.18.04.2
Configuration: /var/www/rector/phpunit.xml

...............................................................  63 / 192 ( 32%)
............................................................... 126 / 192 ( 65%)
............................................................... 189 / 192 ( 98%)
...                                                             192 / 192 (100%)

Time: 9.33 seconds, Memory: 206.00MB

OK (192 tests, 367 assertions)
```

### 2nd run on this PR

It seems PHPStan has some cache for node analysis, that makes 2nd run much faster

```bash
PHPUnit 7.2.6 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.7-0ubuntu0.18.04.2
Configuration: /var/www/rector/phpunit.xml

...............................................................  63 / 192 ( 32%)
............................................................... 126 / 192 ( 65%)
............................................................... 189 / 192 ( 98%)
...                                                             192 / 192 (100%)

Time: 1.96 seconds, Memory: 128.00MB

OK (192 tests, 367 assertions)
```

File: packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php (hunk 3 above)
Comment (carusogabriel) [reply packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php]: It’s commented.

Comment (TomasVotruba) [reply packages/NodeTypeResolver/tests/PerNodeTypeResolver/ClassTypeResolver/ClassTypeResolverTest.php]: Oh, you mean just the part, not the whole test.

Since it's edge case PHPStan skips, I think we can remove that part. Will do

**Comment by TomasVotruba** [comment]
Ready to merge :100:

**Comment by carusogabriel** [comment]
@TomasVotruba Just give me some hours, I wanna to test it in a real application to give us some benchmarks for posterity!

**Comment by TomasVotruba** [comment]
The branch is still available after merge, so feel free to fetch it.
I'm going to merge it so we can continue with clear plate.

**Comment by carusogabriel** [comment]
@TomasVotruba There was no need for the rush, but ok, I'll update with the benchmark later.

**Comment by TomasVotruba** [comment]
@carusogabriel Well, I'd go for benchmark after #553, it will have better impact
